python/compress-image/test_main.py

```diff
@@ -0,0 +1,359 @@
+            # Check if the assert equals and is correct
+            self.assertEqual(optimized_image, (mock_from_buffer.return_value.to_buffer.return_value))
+
+    @unittest.skipIf(not secret.API_KEY_TINYPNG, "No Tinypng API Key set")
+    def test_tinypng_impl_unexpected_exception_account_error(self):
+        """Test case handling unexpected 'AccountError' in 'tinypng_impl'"""
+        with patch("main.tinify.from_buffer") as mock_from_buffer:
+            # Set up the mock return value as account exception
+            mock_from_buffer.side_effect = \
+                tinify.errors.AccountError("API Key is wrong")
+            # Check the raise for Account error
+            self.assertRaises(tinify.errors.AccountError,
+                              main.tinypng_impl,
+                              {"api_key": secret.API_KEY_TINYPNG,
+                               "decoded_image": IMAGE_1KB})
+            mock_from_buffer.assert_called_once()
+
+    @unittest.skipIf(not secret.API_KEY_TINYPNG, "No Tinypng API Key set")
+    def test_tinypng_impl_unexpected_exception_client_error(self):
+        """Test case handling unexpected 'ClientError' in 'tinypng_impl'."""
+        with patch("main.tinify.from_buffer") as mock_from_buffer:
+            # Set up the mock return value as client exception
+            mock_from_buffer.side_effect = \
+                tinify.errors.ClientError("Image is incorrect")
+            # Check the raise for client error
+            self.assertRaises(tinify.errors.ClientError,
+                              main.tinypng_impl,
+                              {"api_key": secret.API_KEY_TINYPNG,
+                               "decoded_image": IMAGE_1KB})
+            mock_from_buffer.assert_called_once()
+
+
+class TestKrakenIO(unittest.TestCase):
+    @unittest.skipIf(not (secret.API_KEY_KRAKENIO and secret.SECRET_API_KEY_KRAKENIO), "No KrakenIO API Key or Secret Key")
+    def test_krakenio(self):
+        """Output validation 1KB"""
+        want = (pathlib.Path(
+            secret.RESULT_1KB_KRAKENIO).read_text(encoding="utf-8"))
+        got = main.krakenio_impl({
+            "api_key": secret.API_KEY_KRAKENIO,
+            "api_secret_key": secret.SECRET_API_KEY_KRAKENIO,
+            "decoded_image": IMAGE_1KB
+        })
+        self.assertEqual(got, base64.b64decode(want))
+
+    @unittest.skipIf(not (secret.API_KEY_KRAKENIO and secret.SECRET_API_KEY_KRAKENIO), "No KrakenIO API Key or Secret Key")
+    def test_krakenio_time_out(self):
+        with patch("main.requests.post") as mock_post:
+            mock_post.side_effect = requests.exceptions.ReadTimeout
+            with self.assertRaises(requests.exceptions.ReadTimeout):
+                main.krakenio_impl({
+                    "api_key": secret.API_KEY_KRAKENIO,
+                    "api_secret_key": secret.SECRET_API_KEY_KRAKENIO,
+                    "decoded_image": IMAGE_1KB
+                })
+            mock_post.assert_called_once()
```

Comment:
If you wanted, you use use `mock.post_assert_called_once_with()` to validate _how_ `post` was called. However, this isn't needed. 

Does it matter if `post()` was called once or twice?